base/applications/taskmgr/CMakeLists.txt

```diff
@@ -24,6 +24,6 @@ file(GLOB taskmgr_rc_deps res/*.*)
 add_rc_deps(taskmgr.rc ${taskmgr_rc_deps})
 add_executable(taskmgr ${SOURCE} taskmgr.rc)
 set_module_type(taskmgr win32gui UNICODE)
-add_importlibs(taskmgr advapi32 user32 gdi32 shell32 shlwapi comctl32 msvcrt kernel32 ntdll)
+add_importlibs(taskmgr advapi32 user32 gdi32 shell32 shlwapi comctl32 ole32 msvcrt kernel32 ntdll)
```

Comment:
ole32? What is it used for? Any alternatives?

Reply:
> ole32? What is it used for? Any alternatives?

For `SHOpenFolderAndSelectItems`

[https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shopenfolderandselectitems](url)

An alternative maybe `ShellExecute`

Reply:
Taskmgr should only use functions from kernel32/ntdll for retrieving a suitable Win32 path, and then *dynamically load* shell32 to use that API.

Reply:
> Taskmgr should only use functions from kernel32/ntdll for retrieving a suitable Win32 path, and then _dynamically load_ shell32 to use that API.

Should I use `CreateProcess` instead of `ShellExecute` ?

Reply:
> An alternative maybe using ShellExecute to start explorer.exe with /select,"<EXE_PATH>" option.

What does Windows' taskmgr do there? (you can see what it does by monitoring it with e.g. ApiMonitor from Rohitab)

Reply:
taskmgr of XPSP3 has no static imports to OLE32 (only dynamic). But it does have static imports to shell32.
So I do vote for hbeluscas proposal to prefer "ShellExectute /select" over adding the OLE32 import
![image](https://user-images.githubusercontent.com/33393466/151670603-42637d93-b0bc-4561-86be-2b0ebfad1531.png)

Reply:
Note that taskmgr has a *delayed-import* to ole32 though.

Reply:
> > An alternative maybe using ShellExecute to start explorer.exe with /select,"<EXE_PATH>" option.
> 
> What does Windows' taskmgr do there? (you can see what it does by monitoring it with e.g. ApiMonitor from Rohitab)

I used Dependency Walker to inspect `taskmgr.exe` from Windows 7, I saw
- ole32.dll and shell32.dll imported
- `SHOpenFolderAndSelectItems`, `ShellExecuteExW` and more shell functions used
- `QueryFullProcessImageNameW` function used

but according to @sashaqwert , Windows 2003 doesn't have this functionality.

Reply:
Do you have a fallback for GetModuleFileNameExW ? (because if so this means that using that function was not necessary to begin with)

Reply:
> Do you have a fallback for GetModuleFileNameExW ? (because if so this means that using that function was not necessary to begin with)

Maybe `NtQueryInformationProcess` but it returns path in device form and requires conversion.

Reply:
> Note that taskmgr has a _delayed-import_ to ole32 though. But I would prefer if all that stuff could be even manually loaded so that one could use taskmgr in an extreme minimal environment that does not contain any shell stuff.

I think we can simply disable this functionality in that case, because we can't open explorer or open properties dialog without shell stuffs anyway.

Reply:
Yes: "disabling the feature whenever delay-load is not found", is a good idea to have good user experience even in modded setups.
Still the shell32-way ShellExecuteExW() is better than the ole32-way for several reasons:
1.) We have a dependency to shell32 already, while we have none to ole32.
2.) it keeps the set of required dlls for full functionality closer to XP/2k3
3.) imho it is more likely to find a "super-slimified" version of ros without ole32, than one without shell32.

Reply:
@JoachimHenze I agree with you to remove ole32 dependency for portability, but I have a question.
Are there any ways to open file properties dialog that work on ROS and don't require ole32? I tried  `ShellExecuteEx` method but it doesn't work for me.

Reply:
Maybe you did wrong with the flags: I just googled, but did not try it:

SHELLEXECUTEINFO ShExecInfo ={0};
ShExecInfo.cbSize = sizeof(SHELLEXECUTEINFO);
ShExecInfo.fMask = SEE_MASK_INVOKEIDLIST ;
ShExecInfo.hwnd = NULL;
ShExecInfo.lpVerb = "properties";
ShExecInfo.lpFile = "c:\\"; //can be a file as well
ShExecInfo.lpParameters = ""; 
ShExecInfo.lpDirectory = NULL;
ShExecInfo.nShow = SW_SHOW;
ShExecInfo.hInstApp = NULL; 
ShellExecuteEx(&ShExecInfo);

above taken from https://www.codeproject.com/Articles/1842/A-newbie-s-elementary-guide-to-spawning-processes
If it wouldn't work out of the box, make sure to try in first when running our taskmgr on Windows. Maybe some other bug in ros (outside of taskmgr) is in your way still.


(I found also something that looks pretty hacky https://stackoverflow.com/questions/33472585/open-windows-file-folder-properties-dialog-from-c (scroll down a bit, the one with that Sleep()...). But that would for sure be no proper solution.

Reply:
> Maybe you did wrong with the flags: I just googled, but did not try it:
> 
> SHELLEXECUTEINFO ShExecInfo ={0}; ShExecInfo.cbSize = sizeof(SHELLEXECUTEINFO); ShExecInfo.fMask = SEE_MASK_INVOKEIDLIST ; ShExecInfo.hwnd = NULL; ShExecInfo.lpVerb = "properties"; ShExecInfo.lpFile = "c:\"; //can be a file as well ShExecInfo.lpParameters = ""; ShExecInfo.lpDirectory = NULL; ShExecInfo.nShow = SW_SHOW; ShExecInfo.hInstApp = NULL; ShellExecuteEx(&ShExecInfo);
> 
> above taken from https://www.codeproject.com/Articles/1842/A-newbie-s-elementary-guide-to-spawning-processes If it wouldn't work out of the box in ros, make sure to try it first when running our taskmgr on Windows. Maybe some other bug in ros (outside of taskmgr) is in your way still.
> 
> (I found also something that looks pretty hacky https://stackoverflow.com/questions/33472585/open-windows-file-folder-properties-dialog-from-c (scroll down a bit, the one with that Sleep()...). But that would for sure be no proper solution.

Thanks. I'll try it.

Reply:
Why the obsession with ole32?
Just make it a delay-import and call it a day, no need to make tons of unreadable code / hacks to prevent loading a system dll that most likely will be loaded by shell32 anyway...

Reply:
shell32 on XP also loads ole32, but it **delay-loads** it. So there is at least a possibility, that shell32 can offer that specific functionality (file prop dlg) without ole32.
But yes: Generally the delay-loading at all is the more important part of the two. The question whether shell32 or ole32 is not **that** important. But still worth a thought, given that shell32 is already a dependency here historically, but ole32 not directly.